pkg/agent/apiserver/handlers/fqdncache/handler_test.go

```diff
@@ -0,0 +1,152 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fqdncache
+
+import (
+	"encoding/json"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"regexp"
+	"regexp/syntax"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+
+	"antrea.io/antrea/pkg/agent/apis"
+	"antrea.io/antrea/pkg/agent/types"
+	"antrea.io/antrea/pkg/querier"
+	queriertest "antrea.io/antrea/pkg/querier/testing"
+)
+
+func TestFqdnCacheQuery(t *testing.T) {
+	expirationTime := time.Now().Add(1 * time.Hour).UTC()
+	tests := []struct {
+		name                 string
+		filteredCacheEntries []types.DnsCacheEntry
+		expectedResponse     []apis.FQDNCacheResponse
+	}{
+		{
+			name: "FQDN cache exists - multiple addresses multiple domains",
+			filteredCacheEntries: []types.DnsCacheEntry{
+				{
+					FQDNName:       "example.com",
+					IPAddress:      net.ParseIP("10.0.0.1"),
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "foo.com",
+					IPAddress:      net.ParseIP("10.0.0.4"),
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "bar.com",
+					IPAddress:      net.ParseIP("10.0.0.5"),
+					ExpirationTime: expirationTime,
+				},
+			},
+			expectedResponse: []apis.FQDNCacheResponse{
+				{
+					FQDNName:       "example.com",
+					IPAddress:      "10.0.0.1",
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "foo.com",
+					IPAddress:      "10.0.0.4",
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "bar.com",
+					IPAddress:      "10.0.0.5",
+					ExpirationTime: expirationTime,
+				},
+			},
+		},
+		{
+			name:                 "FQDN cache does not exist",
+			filteredCacheEntries: []types.DnsCacheEntry{},
+			expectedResponse:     []apis.FQDNCacheResponse{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			q := queriertest.NewMockAgentNetworkPolicyInfoQuerier(ctrl)
+			q.EXPECT().GetFQDNCache(nil).Return(tt.filteredCacheEntries)
+			handler := HandleFunc(q)
+			req, err := http.NewRequest(http.MethodGet, "", nil)
+			require.NoError(t, err)
+			recorder := httptest.NewRecorder()
+			handler.ServeHTTP(recorder, req)
+			var receivedResponse []apis.FQDNCacheResponse
+			err = json.Unmarshal(recorder.Body.Bytes(), &receivedResponse)
+			require.NoError(t, err)
+			assert.Equal(t, tt.expectedResponse, receivedResponse)
+		})
+	}
+}
+
+func TestNewFilterFromURLQuery(t *testing.T) {
+	tests := []struct {
+		name           string
+		queryParams    url.Values
+		expectedFilter *querier.FQDNCacheFilter
+		expectedError  error
+	}{
+		{
+			name:           "Empty query",
+			queryParams:    url.Values{},
+			expectedFilter: nil,
+			expectedError:  nil,
```

Comment:
For conciseness, I would suggest omitting the `expectedError` field when it is `nil`